packages/xchain-haven/src/haven/haven-core-client.ts

```diff
@@ -0,0 +1,319 @@
+  }
+
+  constructor() {
+    this.syncHandler = new SyncHandler()
+  }
+
+  async init(seed: string, netType: string | number): Promise<boolean> {
+    //this.netTypeId = netTypePromise<boolean> {
+    // login and fire up keep_alive
+    this.purge()
+
+    this.netTypeId = typeof netType === 'number' ? netType : (NetTypes[netType as keyof typeof NetTypes] as number)
+    this.seed = seed
+    const apiUrl = this.netTypeId === NetTypes.mainnet ? MainnetApiUrl : TestNetApiUrl
+    setAPI_URL(apiUrl)
+    const keys = this.getKeys()
+
+    setCredentials(keys.address_string, keys.sec_viewKey_string)
+
+    await login(true)
+
+    this.pingServerIntervalID = setInterval(() => this.pingServer(), 60 * 1000)
+
+    return true
+  }
+
+  purge() {
+    this.netTypeId = undefined
+    this.seed = undefined
+    this.syncHandler.purge()
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    //@ts-ignore
+    clearInterval(this.pingServerIntervalID)
+  }
+
+  getAddress(): string {
+    const keys = this.getKeys()
+    return keys.address_string
+  }
+
+  validateAddress(address: string): boolean {
+    const module = this.getCoreModule()
+    let response: string | Record<string, unknown>
+    try {
+      response = module.decode_address(address, this.netTypeId!)
+    } catch (e) {
+      return false
+    }
+    return response.hasOwnProperty('spend')
+  }
+
+  subscribeSyncProgress = async (observer: SyncObserver): Promise<void> =>
+    this.syncHandler.subscribeSyncProgress(observer)
+  getSyncState = async (): Promise<SyncStats> => this.syncHandler.getSyncState()
+  isSyncing = async (): Promise<boolean> => this.syncHandler.isSyncing()
+
+  async getBalance(): Promise<HavenBalance> {
+    const coreModule = this.getCoreModule()
+    const keys = this.getKeys()
+    const { sec_viewKey_string, address_string, pub_spendKey_string, sec_spendKey_string } = keys
+
+    const rawAddressData = await getAddressInfo()
+
+    const serializedData = havenWallet.api_response_parser_utils.Parsed_AddressInfo__sync__keyImageManaged(
+      rawAddressData,
+      address_string,
+      sec_viewKey_string,
+      pub_spendKey_string,
+      sec_spendKey_string,
+      coreModule,
+    )
+
+    const havenBalance: HavenBalance = {} as HavenBalance
+
+    const { total_received_String, total_sent_String, total_received_unlocked_String } = serializedData
+
+    Object.keys(serializedData.total_received_String).forEach((assetType) => {
+      const balance = havenWallet
+        .JSBigInt(total_received_String[assetType as HavenTicker])
+        .subtract(havenWallet.JSBigInt(total_sent_String[assetType as HavenTicker]))
+
+      const unlockedBalance = havenWallet
+        .JSBigInt(total_received_unlocked_String[assetType as HavenTicker])
+        .subtract(havenWallet.JSBigInt(total_sent_String[assetType as HavenTicker]))
+
+      const lockedBalance = balance.subtract(unlockedBalance)
+
+      havenBalance[assetType as HavenTicker] = {
```

Comment:
Try to avoid `assetType as HavenTicker` (see previous comments to similar topic)